File: objc/objcsearch/Sources/objcsearch/SearchResult.m
#import "SearchResult.h"
#import "color.h"

@implementation SearchResult

- (instancetype) initWithPattern:(NSString*)pattern
                            file:(FileResult*)file
                         lineNum:(unsigned long)lineNum
                 matchStartIndex:(unsigned long)matchStartIndex
                   matchEndIndex:(unsigned long)matchEndIndex
                            line:(NSString*)line
                     linesBefore:(NSArray<NSString*>*)linesBefore
                      linesAfter:(NSArray<NSString*>*)linesAfter
{
    self = [super init];
    if (self) {
        self.searchPattern = pattern;
        self.file = file;
        self.lineNum = lineNum;
        self.matchStartIndex = matchStartIndex;
        self.matchEndIndex = matchEndIndex;
        self.line = line;
        self.linesBefore = linesBefore;
        self.linesAfter = linesAfter;
    }
    return self;
}

- (NSString *) getFilePath {
    if (self.file != nil) {
        return [self.file description];
    } else {
        return @"<text>";
    }
}

- (NSComparisonResult)compareBySearchFields:(SearchResult *)otherSearchResult {
    if (self.lineNum == otherSearchResult.lineNum) {
        if (self.matchStartIndex == otherSearchResult.matchStartIndex) {
            if (self.matchEndIndex == otherSearchResult.matchEndIndex) {
                return NSOrderedSame;
            }
            if (self.matchEndIndex < otherSearchResult.matchEndIndex) {
                return NSOrderedAscending;
            }
            return NSOrderedDescending;
        }
        if (self.matchStartIndex < otherSearchResult.matchStartIndex) {
            return NSOrderedAscending;
        }
        return NSOrderedDescending;
    }
    if (self.lineNum < otherSearchResult.lineNum) {
        return NSOrderedAscending;
    }
    return NSOrderedDescending;
}

- (NSComparisonResult)compareByPath:(SearchResult *)otherSearchResult caseInsensitive:(BOOL)caseInsensitive {
    if (self.file != nil && otherSearchResult.file != nil) {
        NSComparisonResult res = [self.file compareByPath:otherSearchResult.file caseInsensitive:caseInsensitive];
        if (res != NSOrderedSame) return res;
    }
    return [self compareBySearchFields:otherSearchResult];
}

- (NSComparisonResult)compareByName:(SearchResult *)otherSearchResult caseInsensitive:(BOOL)caseInsensitive {
    if (self.file != nil && otherSearchResult.file != nil) {
        NSComparisonResult res = [self.file compareByName:otherSearchResult.file caseInsensitive:caseInsensitive];
        if (res != NSOrderedSame) return res;
    }
    return [self compareBySearchFields:otherSearchResult];
}

- (NSComparisonResult)compareBySize:(SearchResult *)otherSearchResult caseInsensitive:(BOOL)caseInsensitive {
    if (self.file != nil && otherSearchResult.file != nil) {
        NSComparisonResult res = [self.file compareBySize:otherSearchResult.file caseInsensitive:caseInsensitive];
        if (res != NSOrderedSame) return res;
    }
    return [self compareBySearchFields:otherSearchResult];
}

- (NSComparisonResult)compareByType:(SearchResult *)otherSearchResult caseInsensitive:(BOOL)caseInsensitive {
    if (self.file != nil && otherSearchResult.file != nil) {
        NSComparisonResult res = [self.file compareByType:otherSearchResult.file caseInsensitive:caseInsensitive];
        if (res != NSOrderedSame) return res;
    }
    return [self compareBySearchFields:otherSearchResult];
}

- (NSComparisonResult)compareByLastMod:(SearchResult *)otherSearchResult caseInsensitive:(BOOL)caseInsensitive {
    if (self.file != nil && otherSearchResult.file != nil) {
        NSComparisonResult res = [self.file compareByLastMod:otherSearchResult.file caseInsensitive:caseInsensitive];
        if (res != NSOrderedSame) return res;
    }
    return [self compareBySearchFields:otherSearchResult];
}

@end

File: csharp/CsSearch/CsSearchLib/SearchResult.cs
using System.Collections.Generic;
using System.Text.RegularExpressions;
using System;

using CsFindLib;

namespace CsSearchLib;

public class SearchResult(
	Regex searchPattern,
	FileResult? file,
	int lineNum,
	int matchStartIndex,
	int matchEndIndex,
	string? line,
	IList<string> linesBefore,
	IList<string> linesAfter)
{
	public Regex SearchPattern { get; private set; } = searchPattern;
	public FileResult? File { get; set; } = file;
	public int LineNum { get; } = lineNum;
	public int MatchStartIndex { get; } = matchStartIndex;
	public int MatchEndIndex { get; private set; } = matchEndIndex;
	public string? Line { get; private set; } = line;
	public IList<string> LinesBefore { get; private set; } = linesBefore;
	public IList<string> LinesAfter { get; private set; } = linesAfter;

	public SearchResult(Regex searchPattern, FileResult? file, int lineNum,
		int matchStartIndex, int matchEndIndex, string? line)
		: this(searchPattern, file, lineNum, matchStartIndex, matchEndIndex, line,
			new List<string>(), new List<string>())
	{
	}

	private int CompareByMatchLocation(SearchResult other)
	{
		if (LineNum != other.LineNum) 
			return LineNum -  other.LineNum;
		return MatchStartIndex - other.MatchStartIndex;
	}

	public int CompareByPath(SearchResult other, bool caseInsensitive)
	{
		if (File != null && other.File != null)
		{
			var cmp = File.CompareByPath(other.File, caseInsensitive);
			if (cmp != 0)
				return cmp;
		}

		return CompareByMatchLocation(other);
	}

	public int CompareByName(SearchResult other, bool caseInsensitive)
	{
		if (File != null && other.File != null)
		{
			var cmp = File.CompareByName(other.File, caseInsensitive);
			if (cmp != 0)
				return cmp;
		}

		return CompareByMatchLocation(other);
	}

	public int CompareBySize(SearchResult other, bool caseInsensitive)
	{
		if (File != null && other.File != null)
		{
			var cmp = File.CompareBySize(other.File, caseInsensitive);
			if (cmp != 0)
				return cmp;
		}

		return CompareByMatchLocation(other);
	}

	public int CompareByType(SearchResult other, bool caseInsensitive)
	{
		if (File != null && other.File != null)
		{
			var cmp = File.CompareByType(other.File, caseInsensitive);
			if (cmp != 0)
				return cmp;
		}

		return CompareByMatchLocation(other);
	}

	public int CompareByLastMod(SearchResult other, bool caseInsensitive)
	{
		if (File != null && other.File != null)
		{
			var cmp = File.CompareByLastMod(other.File, caseInsensitive);
			if (cmp != 0)
				return cmp;
		}

		return CompareByMatchLocation(other);
	}
}
